File: src/engine/linalg.ts
// ─────────────────────────────────────────────────────────────────────────────
//  경량 선형대수 유틸 (소규모 밀집 행렬용)
// ─────────────────────────────────────────────────────────────────────────────

export type Matrix = number[][];

/** A·x = b 를 부분 피벗 가우스 소거법으로 해석 */
export function solveLinear(A: Matrix, b: number[]): number[] {
  const n = b.length;
  const M = A.map((row, i) => [...row, b[i]]);

  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivot][col])) pivot = r;
    }
    if (Math.abs(M[pivot][col]) < 1e-13) {
      throw new Error('Singular matrix in solveLinear');
    }
    [M[col], M[pivot]] = [M[pivot], M[col]];

    const pv = M[col][col];
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const f = M[r][col] / pv;
      if (f === 0) continue;
      for (let c = col; c <= n; c++) M[r][c] -= f * M[col][c];
    }
  }
  return M.map((row, i) => row[n] / row[i]);
}

export function zeros(rows: number, cols: number): Matrix {
  return Array.from({ length: rows }, () => new Array(cols).fill(0));
}

export function transpose(A: Matrix): Matrix {
  const rows = A.length;
  const cols = A[0]?.length ?? 0;
  const T = zeros(cols, rows);
  for (let i = 0; i < rows; i++)
    for (let j = 0; j < cols; j++) T[j][i] = A[i][j];
  return T;
}

export function matMul(A: Matrix, B: Matrix): Matrix {
  const n = A.length;
  const m = B[0].length;
  const k = B.length;
  const C = zeros(n, m);
  for (let i = 0; i < n; i++) {
    for (let p = 0; p < k; p++) {
      const a = A[i][p];
      if (a === 0) continue;
      for (let j = 0; j < m; j++) C[i][j] += a * B[p][j];
    }
  }
  return C;
}

export function matVec(A: Matrix, x: number[]): number[] {
  return A.map((row) => row.reduce((s, v, j) => s + v * x[j], 0));
}

/** 대칭 양정치(가정) 정방행렬의 역행렬 (가우스-조던) */
export function invert(A: Matrix): Matrix {
  const n = A.length;
  const M = A.map((row, i) => [
    ...row,
    ...Array.from({ length: n }, (_, j) => (i === j ? 1 : 0)),
  ]);

  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivot][col])) pivot = r;
    }
    if (Math.abs(M[pivot][col]) < 1e-13) {
      throw new Error('Singular matrix in invert');
    }
    [M[col], M[pivot]] = [M[pivot], M[col]];

    const pv = M[col][col];
    for (let c = 0; c < 2 * n; c++) M[col][c] /= pv;

    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const f = M[r][col];
      if (f === 0) continue;
      for (let c = 0; c < 2 * n; c++) M[r][c] -= f * M[col][c];
    }
  }
  return M.map((row) => row.slice(n));
}

File: python/linalg.py
"""
선형 연립방정식 풀이 — 가우스 소거법 (Gaussian Elimination)
외부 라이브러리(numpy) 없이 직접 구현
"""


def solve_linear(A: list, b: list) -> list:
    """
    Ax = b 풀기 (가우스 소거 + 후진 대입)
    A: n×n 행렬, b: 길이 n 벡터
    반환: x (길이 n 벡터)
    """
    n = len(b)
    # 증강행렬 구성
    M = [A[i][:] + [b[i]] for i in range(n)]

    # 전진 소거
    for col in range(n):
        # 피벗 선택 (부분 피벗팅)
        max_row = max(range(col, n), key=lambda r: abs(M[r][col]))
        M[col], M[max_row] = M[max_row], M[col]
        pivot = M[col][col]
        if abs(pivot) < 1e-15:
            raise ValueError(f"행렬이 특이(singular)합니다. col={col}")
        for row in range(col+1, n):
            factor = M[row][col] / pivot
            for k in range(col, n+1):
                M[row][k] -= factor * M[col][k]

    # 후진 대입
    x = [0.0]*n
    for i in range(n-1, -1, -1):
        x[i] = M[i][n]
        for j in range(i+1, n):
            x[i] -= M[i][j] * x[j]
        x[i] /= M[i][i]

    return x
